File: lib/calc.ts
import { roundHalfUp, fractionalPart, assert, stableSortBy } from "./utils";

export type CalcInput = {
  items: { id: string; priceCents: number; taxable?: boolean; taxRatePct?: number }[];
  participants: { id: string; name: string }[];
  shares: { itemId: string; participantId: string; weight: number }[];
  taxRatePct: number;
  tipRatePct: number;
  taxMode?: "GLOBAL" | "ITEM";
  convenienceFeeRatePct?: number;
};

export type CalcOutput = {
  billTotals: {
    subtotalCents: number;
    taxCents: number;
    tipCents: number;
    convenienceFeeCents: number;
    grandTotalCents: number;
  };
  participants: {
    participantId: string;
    name: string;
    preTaxCents: number;
    taxCents: number;
    tipCents: number;
    totalOwedCents: number;
  }[];
  byItem: {
    itemId: string;
    allocations: { participantId: string; cents: number }[];
  }[];
};

export function calculateSplit(input: CalcInput): CalcOutput {
  const { items, participants, shares, taxRatePct, tipRatePct, taxMode = "GLOBAL", convenienceFeeRatePct = 0 } = input;

  // 1. Subtotal
  const subtotalCents = items.reduce((acc, it) => acc + it.priceCents, 0);

  // 2-3. Item splits with weights, round-half-up per participant and fix rounding drift
  const participantIdToPreTaxCents = new Map<string, number>();
  for (const p of participants) participantIdToPreTaxCents.set(p.id, 0);

  const byItem: { itemId: string; allocations: { participantId: string; cents: number }[] }[] = [];

  for (const item of items) {
    const itemShares = shares.filter((s) => s.itemId === item.id && s.weight > 0);
    if (itemShares.length === 0) {
      throw new Error(`Item "${item.id}" has no shares assigned.`);
    }

    const totalWeight = itemShares.reduce((acc, s) => acc + s.weight, 0);
    // Normalized allocations
    const exactAllocations = itemShares.map((s) => ({
      participantId: s.participantId,
      exact: (item.priceCents * s.weight) / totalWeight,
    }));

    const rounded = exactAllocations.map((ea) => ({
      participantId: ea.participantId,
      exact: ea.exact,
      rounded: roundHalfUp(ea.exact),
      frac: fractionalPart(ea.exact),
    }));

    let sumRounded = rounded.reduce((acc, r) => acc + r.rounded, 0);
    const delta = item.priceCents - sumRounded; // how many pennies to adjust

    if (delta !== 0) {
      const direction = Math.sign(delta); // +1 add pennies, -1 remove pennies
      const count = Math.abs(delta);
      const sorted = rounded
        .slice()
        .sort((a, b) => {
          // For adding pennies: larger fractional parts first
          // For removing pennies: smaller fractional parts first
          const primary = direction > 0 ? b.frac - a.frac : a.frac - b.frac;
          if (primary !== 0) return primary;
          // tiebreak by participantId for determinism
          return a.participantId.localeCompare(b.participantId);
        });
      for (let i = 0; i < count; i++) {
        sorted[i % sorted.length].rounded += direction;
      }
      // Recompute sum after distribution
      sumRounded = sorted.reduce((acc, r) => acc + r.rounded, 0);
      // Place back into rounded map
      const byId = new Map(sorted.map((r) => [r.participantId, r.rounded] as const));
      for (const r of rounded) r.rounded = byId.get(r.participantId) ?? r.rounded;
    }

    // Accumulate pre-tax per participant
    const itemAllocations: { participantId: string; cents: number }[] = [];
    for (const r of rounded) {
      const prev = participantIdToPreTaxCents.get(r.participantId) ?? 0;
      participantIdToPreTaxCents.set(r.participantId, prev + r.rounded);
      itemAllocations.push({ participantId: r.participantId, cents: r.rounded });
    }
    byItem.push({ itemId: item.id, allocations: itemAllocations });
  }

  // 5. Pools
  let taxCents = 0;
  if (taxMode === "GLOBAL") {
    const taxableSubtotal = subtotalCents;
    taxCents = roundHalfUp((taxableSubtotal * taxRatePct) / 100);
  } else {
    // Sum item-level tax based on each item's specific tax rate (or 0)
    taxCents = items.reduce((acc, it) => {
      if (it.taxable === false) return acc;
      const rate = typeof it.taxRatePct === "number" ? it.taxRatePct : taxRatePct;
      return acc + roundHalfUp((it.priceCents * rate) / 100);
    }, 0);
  }
  const tipCents = roundHalfUp((subtotalCents * tipRatePct) / 100);

  const preTaxTotals = participants.map((p) => ({
    participantId: p.id,
    name: p.name,
    preTaxCents: participantIdToPreTaxCents.get(p.id) ?? 0,
  }));

  // 6. Allocate pools proportionally using floor + remainder distribution
  const poolAllocate = (pool: number) => {
    if (pool === 0 || subtotalCents === 0) {
      return preTaxTotals.map(() => 0);
    }
    const exacts = preTaxTotals.map((pt) => (pool * pt.preTaxCents) / subtotalCents);
    const floors = exacts.map((e) => Math.floor(e));
    let sumFloors = floors.reduce((a, b) => a + b, 0);
    let leftover = pool - sumFloors;
    if (leftover > 0) {
      const order = preTaxTotals
        .map((pt, idx) => ({ idx, frac: fractionalPart(exacts[idx]) }))
        .sort((a, b) => b.frac - a.frac || (preTaxTotals[a.idx].participantId.localeCompare(preTaxTotals[b.idx].participantId)));
      for (let i = 0; i < leftover; i++) floors[order[i % order.length].idx] += 1;
      sumFloors = floors.reduce((a, b) => a + b, 0);
    }
    return floors;
  };

  const feeCents = roundHalfUp((subtotalCents * convenienceFeeRatePct) / 100);
  const taxAlloc = poolAllocate(taxCents);
  const tipAlloc = poolAllocate(tipCents);
  const feeAlloc = poolAllocate(feeCents);

  const participantsOut = preTaxTotals.map((pt, idx) => {
    const tax = taxAlloc[idx];
    const tip = tipAlloc[idx];
    const fee = feeAlloc[idx];
    const total = pt.preTaxCents + tax + tip + fee;
    return {
      participantId: pt.participantId,
      name: pt.name,
      preTaxCents: pt.preTaxCents,
      taxCents: tax,
      tipCents: tip,
      convenienceFeeCents: fee,
      totalOwedCents: total,
    };
  });

  // 8. Assert exactness
  const sumOwed = participantsOut.reduce((acc, p) => acc + p.totalOwedCents, 0);
  const grandTotal = subtotalCents + taxCents + tipCents + feeCents;
  assert(sumOwed === grandTotal, "Totals do not add up exactly.");

  return {
    billTotals: {
      subtotalCents,
      taxCents,
      tipCents,
      convenienceFeeCents: feeCents,
      grandTotalCents: grandTotal,
    },
    participants: participantsOut,
    byItem,
  };
}

File: mobile/lib/utils/calc.dart
/// Dart port of `web/lib/calc.ts`.
///
/// Used to preview splits offline (e.g. before a bill is persisted) and to
/// keep the logic available even when the device is offline. Behaviour must
/// stay byte-for-byte equivalent with the canonical TypeScript version:
///
///  - round half up (away from zero)
///  - distribute rounding remainder to participants with the largest fractional
///    part (ties broken by participantId)
///  - allocate tax/tip/fee pools by floor + remainder distribution
///  - assert sums equal the grand total
class CalcItem {
  final String id;
  final int priceCents;
  final bool taxable;
  final double? taxRatePct;
  CalcItem({
    required this.id,
    required this.priceCents,
    this.taxable = true,
    this.taxRatePct,
  });
}

class CalcParticipant {
  final String id;
  final String name;
  CalcParticipant({required this.id, required this.name});
}

class CalcShare {
  final String itemId;
  final String participantId;
  final double weight;
  CalcShare({
    required this.itemId,
    required this.participantId,
    required this.weight,
  });
}

class CalcParticipantResult {
  final String participantId;
  final String name;
  final int preTaxCents;
  final int taxCents;
  final int tipCents;
  final int convenienceFeeCents;
  final int totalOwedCents;
  CalcParticipantResult({
    required this.participantId,
    required this.name,
    required this.preTaxCents,
    required this.taxCents,
    required this.tipCents,
    required this.convenienceFeeCents,
    required this.totalOwedCents,
  });
}

class CalcTotals {
  final int subtotalCents;
  final int taxCents;
  final int tipCents;
  final int convenienceFeeCents;
  final int grandTotalCents;
  CalcTotals({
    required this.subtotalCents,
    required this.taxCents,
    required this.tipCents,
    required this.convenienceFeeCents,
    required this.grandTotalCents,
  });
}

class CalcOutput {
  final CalcTotals billTotals;
  final List<CalcParticipantResult> participants;
  CalcOutput({required this.billTotals, required this.participants});
}

int roundHalfUp(double value) {
  if (value.isNaN || !value.isFinite) return 0;
  return value >= 0 ? (value + 0.5).floor() : (value - 0.5).ceil();
}

double fractionalPart(double value) => value - value.floor();

CalcOutput calculateSplit({
  required List<CalcItem> items,
  required List<CalcParticipant> participants,
  required List<CalcShare> shares,
  required double taxRatePct,
  required double tipRatePct,
  String taxMode = 'GLOBAL',
  double convenienceFeeRatePct = 0,
}) {
  final subtotalCents = items.fold<int>(0, (a, it) => a + it.priceCents);

  final preTax = <String, int>{
    for (final p in participants) p.id: 0,
  };

  for (final item in items) {
    final itemShares = shares
        .where((s) => s.itemId == item.id && s.weight > 0)
        .toList();
    if (itemShares.isEmpty) {
      throw StateError('Item "${item.id}" has no shares assigned.');
    }
    final totalWeight = itemShares.fold<double>(0, (a, s) => a + s.weight);
    final exact = itemShares
        .map((s) => _Alloc(
              participantId: s.participantId,
              exact: (item.priceCents * s.weight) / totalWeight,
            ))
        .toList();
    for (final a in exact) {
      a.rounded = roundHalfUp(a.exact);
      a.frac = fractionalPart(a.exact);
    }
    int sumRounded = exact.fold<int>(0, (a, r) => a + r.rounded);
    final delta = item.priceCents - sumRounded;
    if (delta != 0) {
      final direction = delta.sign;
      final count = delta.abs();
      final sorted = [...exact]..sort((a, b) {
          final primary = direction > 0 ? b.frac - a.frac : a.frac - b.frac;
          if (primary != 0) return primary.compareTo(0) < 0 ? -1 : 1;
          return a.participantId.compareTo(b.participantId);
        });
      for (int i = 0; i < count; i++) {
        sorted[i % sorted.length].rounded += direction;
      }
    }
    for (final a in exact) {
      preTax[a.participantId] = (preTax[a.participantId] ?? 0) + a.rounded;
    }
  }

  int taxCents;
  if (taxMode == 'GLOBAL') {
    taxCents = roundHalfUp((subtotalCents * taxRatePct) / 100);
  } else {
    taxCents = items.fold<int>(0, (a, it) {
      if (!it.taxable) return a;
      final rate = it.taxRatePct ?? taxRatePct;
      return a + roundHalfUp((it.priceCents * rate) / 100);
    });
  }
  final tipCents = roundHalfUp((subtotalCents * tipRatePct) / 100);
  final feeCents = roundHalfUp((subtotalCents * convenienceFeeRatePct) / 100);

  List<int> allocatePool(int pool) {
    if (pool == 0 || subtotalCents == 0) {
      return List<int>.filled(participants.length, 0);
    }
    final exacts = participants
        .map((p) => (pool * (preTax[p.id] ?? 0)) / subtotalCents)
        .toList();
    final floors = exacts.map((e) => e.floor()).toList();
    final sumFloors = floors.fold<int>(0, (a, b) => a + b);
    final leftover = pool - sumFloors;
    if (leftover > 0) {
      final order = List<int>.generate(participants.length, (i) => i)
        ..sort((a, b) {
          final cmp = fractionalPart(exacts[b]).compareTo(fractionalPart(exacts[a]));
          if (cmp != 0) return cmp;
          return participants[a].id.compareTo(participants[b].id);
        });
      for (int i = 0; i < leftover; i++) {
        floors[order[i % order.length]] += 1;
      }
    }
    return floors;
  }

  final taxAlloc = allocatePool(taxCents);
  final tipAlloc = allocatePool(tipCents);
  final feeAlloc = allocatePool(feeCents);

  final out = <CalcParticipantResult>[];
  for (var i = 0; i < participants.length; i++) {
    final p = participants[i];
    final pre = preTax[p.id] ?? 0;
    final t = taxAlloc[i];
    final tip = tipAlloc[i];
    final fee = feeAlloc[i];
    out.add(CalcParticipantResult(
      participantId: p.id,
      name: p.name,
      preTaxCents: pre,
      taxCents: t,
      tipCents: tip,
      convenienceFeeCents: fee,
      totalOwedCents: pre + t + tip + fee,
    ));
  }

  final grand = subtotalCents + taxCents + tipCents + feeCents;
  final sumOwed = out.fold<int>(0, (a, r) => a + r.totalOwedCents);
  if (sumOwed != grand) {
    throw StateError('Totals do not add up exactly.');
  }

  return CalcOutput(
    billTotals: CalcTotals(
      subtotalCents: subtotalCents,
      taxCents: taxCents,
      tipCents: tipCents,
      convenienceFeeCents: feeCents,
      grandTotalCents: grand,
    ),
    participants: out,
  );
}

class _Alloc {
  final String participantId;
  final double exact;
  int rounded = 0;
  double frac = 0;
  _Alloc({required this.participantId, required this.exact});
}
